### ISSUE TITLE
Concurrency alongside DBTransactionQueue

### ISSUE THREAD

**Issue by mblcdr** [Raizlabs/DBFlow#483]
I am using DBFlow to show suggestions in a AutoCompleteTextView on android. All of my inserts and updates and deletes go through DBTransactionQueue using custom transactions.
As you may know AutoCompleteTextView uses a Filterable Adapter which does its filtering in a worker thread (naturally) and provides the results on UI thread.
As the filtering (and so select query on database) happens in a worker thread it has to be a blocking operation, and can not go through the DBTransactionQueue which provide the results asynchronously (Unless there is a way to do blocking work with DBTransactionQueue which I am not aware of).
I was using a simple select query for filtering and encountered some exceptions like the one below:

```Java
Fatal Exception: java.lang.RuntimeException: android.database.sqlite.SQLiteDatabaseLockedException: database is locked (code 5): retrycount exceeded
       at com.e.a.a.d.b.run(DBTransactionQueue.java:75)
Caused by android.database.sqlite.SQLiteDatabaseLockedException: database is locked (code 5): retrycount exceeded
       at android.database.sqlite.SQLiteConnection.nativeExecuteForCursorWindow(SQLiteConnection.java)
       at android.database.sqlite.SQLiteConnection.executeForCursorWindow(SQLiteConnection.java:843)
       at android.database.sqlite.SQLiteSession.executeForCursorWindow(SQLiteSession.java:836)
       at android.database.sqlite.SQLiteQuery.fillWindow(SQLiteQuery.java:62)
       at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:143)
       at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:132)
       at android.database.AbstractCursor.moveToPosition(AbstractCursor.java:219)
       at android.database.AbstractCursor.moveToFirst(AbstractCursor.java:258)
       at com.e.a.a.e.c.a(SqlUtils.java:161)
       at com.e.a.a.e.c.b(SqlUtils.java:224)
       at com.e.a.a.e.b.a.c(BaseModelQueriable.java:40)
       at com.e.a.a.e.b.k.c(Where.java:366)
       at com.nazdika.app.c.h.a(SuggestedUserListTransaction.java:40)
       at com.nazdika.app.c.h.a(SuggestedUserListTransaction.java:22)
       at com.e.a.a.d.a.a.f.a(ProcessModelTransaction.java:76)
       at com.e.a.a.d.a.a.c.run(ProcessModelHelper.java:33)
       at com.e.a.a.d.g.a(TransactionManager.java:108)
       at com.e.a.a.d.a.a.b.a(ProcessModelHelper.java:28)
       at com.e.a.a.d.a.a.d.a(ProcessModelInfo.java:159)
       at com.e.a.a.d.a.a.e.c(ProcessModelTransaction.java:73)
       at com.e.a.a.d.a.a.e.a(ProcessModelTransaction.java:13)
       at com.e.a.a.d.b.run(DBTransactionQueue.java:61)
```

So as this is a common use case of databases in android, is there a way around this issue? Maybe a way to stop DBTransactionQueue form executing transactions until the filtering is over?

**Comment by agrosner** [comment]
have you tried making the query just synchronous since the completion happens on a worker thread, and using an `Index` combined with Model caching for super-fast lookup?

**Comment by mblcdr** [comment]
Thank you for taking time to help with the problem.

By synchronous, do you mean in UI thread? That would not solve the issue. Anyhow it is accessing the db from multiple threads. One from the DBTransactionQueue (which is saving models in its worker thread) and one in UI or other background thread.
What I don't understand is why select (which doesn't write to DB) is getting locked. Here's my processModel method, maybe there's problem with it:

```Java
@Override
public void processModel(Model model) {
    try {
        SuggestedUser user = new Select().from(SuggestedUser.class)
                .byIds(model.id).querySingle();

        if (user == null) {
            model.entryCount = Values.assignEntryCount();
            model.hits = 1;
            model.save();
        } else {
            user.hits++;
            user.fillData(model);
            user.update();
        }
    } catch (Throwable t) {
        t.printStackTrace();
    }
}
```

**Comment by agrosner** [comment]
should be fixed since 3.0.0-beta5 with overhauled transaction system.
